benchmarks/benchmarks/benchmark_dataset.py

```diff
@@ -0,0 +1,164 @@
+        self._td.cleanup()
+
+    def time_many_small_reads(self):
+        ds = self.f['a']
+        for i in range(10000):
+            arr = ds[i * 10:(i + 1) * 10]
+
+class TimeWritingSuite:
+    """Based on example in GitHub issue 492:
+    https://github.com/h5py/h5py/issues/492
+    """
+    def setup(self):
+        self._td = TemporaryDirectory()
+        path = osp.join(self._td.name, 'test.h5')
+        self.f = h5py.File(path, 'w')
+        self.shape = shape = (128, 1024, 512)
+        self.f.create_dataset(
+            'a', shape=shape, dtype=np.float32, chunks=(1, shape[1], 64)
+        )
+        self.ds = self.f['a']
+        self.data = np.zeros(self.shape[:2])
+        self.datax = np.zeros(self.shape[1:])
+
+    def teardown(self):
+        self.f.close()
+        self._td.cleanup()
+
+    def time_write_index_first_axis(self):
+        for i in range(self.shape[0]):
+            self.ds[i,:,:] = self.datax
+
+    def time_write_slice_first_axis(self):
+        for i in range(self.shape[0]):
+            self.ds[i:i+1,:,:] = self.datax[np.newaxis, ...]
+
+    def time_write_index_last_axis(self):
+        for i in range(self.shape[2]):
+            self.ds[..., i] = self.data
+
+    def time_write_slice_last_axis(self):
+        for i in range(self.shape[2]):
+            self.ds[..., i:i+1] = self.data[..., np.newaxis]
+
+
+class TimeDatasetSuite:
+
+    def setup(self):
+        self._td = TemporaryDirectory()
+        self.path = osp.join(self._td.name, '_dataset.h5')
+        self.f = h5py.File(self.path, 'w')
+
+    def teardown(self):
+        self.f.close()
+        self._td.cleanup()
+
+    def time_create_dataset(self):
+        self.f.create_dataset('dataset', (100,), dtype='f')
+
+    def time_create_empty_dataset(self):
+        self.f.create_dataset('empty', dtyp='f')
+
+class TimeSuite:
+
+    def setup(self):
+        self._td = TemporaryDirectory()
+        self.path = osp.join(self._td.name, '_operate.h5')
+        self.f = h5py.File(self.path, 'w')
+
+        self.dset = self.f.create_dataset('dset', (100,))
+        self.arr = np.ones((100,))
+        self.arr2 = np.ones((200,))
+
+    def teardown(self):
+        self.f.close()
+        self._td.cleanup()
+
+    def time_read_direct(self):
+        self.dset.read_direct(self.arr, np.s_[0:10], np.s_[50:60])
+
+    def time_write_direct(self):
+        self.dset.write_direct(self.arr2, np.s_[0:10], np.s_[50:60])
+
+
+class TimeVirtualDataset:
+
+    def setup(self):
+        self.layout = h5py.VirtualLayout(shape=(4,100), dtype='i4')
+        for n in range(4):
+            filename = "{}.h5".format(n)
+            vsource = h5py.VirtualSource(filename, 'data', shape=(100,))
+            self.layout[n] = vsource
+
+        self.f = h5py.File('VDS.h5', 'w', libver='latest')
+
+    def teardown(self):
+        self.f.close()
+
+    def time_virtual_dataset(self):
+        self.f.create_virtual_dataset('vdata', self.layout, fillvalue=-1)
```

Comment:
I haven't looked at all of these in detail, but does it make sense to time only this bit, leaving the layout/source bits in setup? For that matter, does it make sense to benchmark creating virtual datasets at all? I can probably make a good case that it's not particularly important for performance.

The more general point is that we need to design benchmarks that can tell us something interesting, which isn't always easy.